flatfs.go

```diff
@@ -1210,46 +1086,452 @@ func (fs *Datastore) Close() error {
+	if bt.asyncFirstError == nil {
+		bt.asyncFirstError = err
+	}
+}
+
+// getAsyncError returns the first error from an async write operation
+func (bt *flatfsBatch) getAsyncError() error {
+	bt.asyncMu.Lock()
+	defer bt.asyncMu.Unlock()
+
+	err := bt.asyncFirstError
+	return err
+}
+
 func (bt *flatfsBatch) Delete(ctx context.Context, key datastore.Key) error {
+	bt.mu.Lock()
+	defer bt.mu.Unlock()
+
+	if err := bt.getAsyncError(); err != nil {
+		return err
+	}
+
 	if keyIsValid(key) {
 		bt.deletes[key] = struct{}{}
 	} // otherwise, delete is a no-op anyways.
 	return nil
 }
 
+func (bt *flatfsBatch) Get(ctx context.Context, key datastore.Key) ([]byte, error) {
+	// Wait for all async writes to complete before reading
+	bt.asyncWrites.Wait()
+	if err := bt.getAsyncError(); err != nil {
+		return nil, err
+	}
+	bt.mu.Lock()
+	// Check if key is marked for deletion
+	if _, deleted := bt.deletes[key]; deleted {
+		bt.mu.Unlock()
+		return nil, datastore.ErrNotFound
+	}
+
+	// Check if key was added in this batch
+	inBatch := slices.Contains(bt.puts, key)
+	bt.mu.Unlock()
+
+	// If in batch, read from temp directory
+	if inBatch {
+		noslash := key.String()[1:]
+		tempFile := filepath.Join(bt.tempDir, noslash+extension)
+		data, err := readFile(tempFile)
+		if err != nil {
+			return nil, err
+		}
+		return data, nil
+	}
+
+	// If not in batch, check main datastore
+	return bt.ds.Get(ctx, key)
+}
+
+func (bt *flatfsBatch) Has(ctx context.Context, key datastore.Key) (bool, error) {
+	// Wait for all async writes to complete before checking
+	bt.asyncWrites.Wait()
+	if err := bt.getAsyncError(); err != nil {
+		return false, err
+	}
+
+	bt.mu.Lock()
+	// Check if key is marked for deletion
+	if _, deleted := bt.deletes[key]; deleted {
+		bt.mu.Unlock()
+		return false, nil
+	}
+
+	// Check if key was added in this batch
+	inBatch := slices.Contains(bt.puts, key)
+	bt.mu.Unlock()
+
+	if inBatch {
+		return true, nil
+	}
+
+	// If not in batch, check main datastore
+	return bt.ds.Has(ctx, key)
+}
+
+func (bt *flatfsBatch) GetSize(ctx context.Context, key datastore.Key) (int, error) {
+	// Wait for all async writes to complete before checking size
+	bt.asyncWrites.Wait()
+
+	if err := bt.getAsyncError(); err != nil {
+		return 0, err
+	}
+	bt.mu.Lock()
+	// Check if key is marked for deletion
+	if _, deleted := bt.deletes[key]; deleted {
+		bt.mu.Unlock()
+		return -1, datastore.ErrNotFound
+	}
+
+	// Check if key was added in this batch
+	inBatch := slices.Contains(bt.puts, key)
+	bt.mu.Unlock()
+
+	// If in batch, get size from temp directory
+	if inBatch {
+		noslash := key.String()[1:]
+		tempFile := filepath.Join(bt.tempDir, noslash+extension)
+		stat, err := os.Stat(tempFile)
+		if err != nil {
+			return -1, err
+		}
+		return int(stat.Size()), nil
+	}
+
+	// If not in batch, check main datastore
+	return bt.ds.GetSize(ctx, key)
+}
+
+func (bt *flatfsBatch) Query(ctx context.Context, q query.Query) (query.Results, error) {
```

Comment:
Is this accurate description?
```suggestion
// Query returns all entries from both the batch and underlying datastore,
// properly merging results to reflect the batch's uncommitted state.
//
// Merge logic:
// - Keys written via Put appear in results (even if not committed)
// - Keys marked for Delete do not appear in results
// - Keys Put multiple times appear only once (last write wins)
// - Main datastore results are excluded if overwritten or deleted in batch
//
// The implementation waits for all async writes to complete before querying.
func (bt *flatfsBatch) Query(ctx context.Context, q query.Query) (query.Results, error) {
```